New bootstrap: track priority weights rather than backoff weights to reduce memory consumption.

clemahieu opened the pull request:
The purpose of this is to fix memory/space consumption of tracked weights. Accounts implicitly have the lowest priority if they're not stored in memory which gives the greatest memory reduction.

clemahieu committed:
Removing forwarding set as it will be replaced merged in to a priority set.

clemahieu committed:
Converting the ascending bootstrapper from tracking backoff weights to tracking priority weights. The purpose of this is to fix memory/space consumption of tracked weights. Accounts implicitly have the lowest priority if they're not stored in memory which gives the greatest memory reduction.

clemahieu committed:
Evenly splitting prioritized and non-prioritized accounts within exclusion limit.

clemahieu committed:
Factoring priority_new out of conditional logic.

clemahieu committed:
Converting priority container to multi-index and trimming lowest priority account when size exceeds a maximum.

clemahieu committed:
Throttling the ascending bootstrapper to not overload block_processor which might also waste bootstrap requests.

dsiganos commented on nano/node/bootstrap/bootstrap_ascending.hpp:
```diff
@@ -191,7 +210,7 @@ namespace bootstrap
 		account_sets accounts;
 		connection_pool pool;
 
-		static std::size_t constexpr parallelism = 16;
```
Why is parallelism down to 1?

clemahieu replied:
Reinstated.

dsiganos commented on nano/node/bootstrap/bootstrap_ascending.hpp:
```diff
@@ -68,12 +69,13 @@ namespace bootstrap
 			 * If an account is not blocked, then add it to the forwarding set and, additioally, if it does not exist in the backoff set, set the priority.
 			 */
-			void prioritize (nano::account const & account, float priority);
+			void priority_up (nano::account const & account);
```
It would be nice to comment this block of functions:
```
void priority_up (nano::account const & account);
void priority_down (nano::account const & account);
void block (nano::account const & account, nano::block_hash const & dependency);
void unblock (nano::account const & account, nano::block_hash const & hash);
void force_unblock 
 ```

dsiganos commented on nano/node/bootstrap/bootstrap_ascending.cpp:
```diff
@@ -182,40 +205,85 @@ void nano::bootstrap::bootstrap_ascending::account_sets::unblock (nano::account
-	return result;
 }
 
 nano::account nano::bootstrap::bootstrap_ascending::account_sets::random ()
```
It would be nice to explain what this function does.
It is not easy to understand.

clemahieu replied:
Added clarity

dsiganos commented on nano/node/bootstrap/bootstrap_ascending.cpp:
```diff
@@ -226,21 +294,8 @@ nano::account nano::bootstrap::bootstrap_ascending::account_sets::random ()
-		result = random ();
-	}
-	backoff[result] += 1.0f;
+	stats.inc (nano::stat::type::bootstrap_ascending_accounts, nano::stat::detail::next_random);
```
It is a bit surprising that next is equivalent to random unless random does not really mean random in which case maybe it should be renamed or at least documented.

clemahieu replied:
Removed.

dsiganos commented:
Also, we should decide which branch will become the v24 branch because it is starting to look like it will not be branch new-bootstrap and adding work to branch new-bootstrap is likely to be wasted effort.

clemahieu committed:
Commenting multiple interface functions
Removing account_sets::next which was the same as account_sets::random
Reinstated parallelism to 16.

clemahieu committed:
Re enabling limiter so as to not overfill block processor.

dsiganos commented:
This is not relevant any more. Development has moved on, on another branch. Closing.
